crates/turbo-tasks-bytes/src/stream.rs

```diff
@@ -0,0 +1,219 @@
+use std::{
+    fmt, mem,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context as TaskContext, Poll},
+};
+
+use anyhow::Result;
+use futures::{Stream as StreamTrait, StreamExt};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+
+/// Streams allow for streaming values from source to sink.
+///
+/// A Stream implements both a reader (which implements the Stream trait), and a
+/// writer (which can be sent to another thread). As new values are written, any
+/// pending readers will be woken up to receive the new value.
+#[derive(Debug)]
+pub struct Stream<T> {
+    inner: Arc<Mutex<StreamState<T>>>,
+}
+
+/// The StreamState actually holds the data of a Stream.
+pub enum StreamState<T> {
+    /// An OpenStream is tied directly to a source stream, and will lazily pull
+    /// new values out as a reader reaches the end of our already-pulled
+    /// data.
+    OpenStream {
+        source: Box<dyn StreamTrait<Item = T> + Send + Sync + Unpin>,
+        pulled: Vec<T>,
+    },
+
+    /// A Closed stream state cannot be pushed to, so it's anyone polling can
+    /// read all values at their leisure.
+    Closed(Box<[T]>),
+}
```

Comment:
* It's a little bit weird that the pulled data is either in `OpenStream.pulled` or in `Closed.0`. I would hoist that into a shared field.
* The current design doesn't allow to free memory of the stream, even if all reader have read a part of the stream. That doesn't allow to use the stream with infinite stream. That's currently not needed, but would be a cool feature to have.

So I would do it this way

```suggestion
/// The StreamState actually holds the data of a Stream.
pub struct StreamSegment<T> {
    /// Data in the current segment. More can be pushed into it as it's pulled from the source.
    /// When this reaches a limit, it will create a new segment into `next` rather than pushing more
    /// items to the buffer. This allows to deallocate previous segements once all reader have consumed them.
    buffer: Vec<T>,

    /// What follows after this segment.
    /// Might be `None` when the source is ended/closed.
    next: Option<StreamSegmentNext<T>>,
}

enum StreamSegmentNext<T> {
    /// A stream is tied directly to a source stream, and will lazily pull
    /// new values out as a reader reaches the end of our already-pulled
    /// data.
    Source(Box<dyn StreamTrait<Item = T> + Send + Sync + Unpin>),
    /// Another segment is following. This forms a linked list of segments to
    /// allow deallocating memory of previous segments.
    Segment(Arc<Mutex<StreamSegment<T>>>
}
```

Should streams have a notation of "errored"? If so we might want to add `StreamSegementNext::Error(SharedError)`?

Reply:
> The current design doesn't allow to free memory of the stream, even if all reader have read a part of the stream. That doesn't allow to use the stream with infinite stream. That's currently not needed, but would be a cool feature to have.

That would mean the Vc which holds a stream could never be reused, and any tasks which depend on it need to rerun the node evaluation fully.

I can make join the open/closed stream into the same struct by wrapping the `source` in an `Option<dyn StreamTrait>`.

> Should streams have a notation of "errored"? If so we might want to add StreamSegementNext::Error(SharedError)?

`Stream` is generic, the wrappers can (and do) handle error state with `Result<T>` inside the stream.